lib/sqlbuilder/batch.go

```diff
@@ -19,51 +14,35 @@ func newBatchInserter(inserter *inserter, size int) *BatchInserter {
 	b := &BatchInserter{
 		inserter: inserter,
 		size:     size,
-		next:     make(chan Inserter),
+		Values:   make(chan interface{}, size),
 	}
-	b.reset()
 	return b
 }
 
-func (b *BatchInserter) reset() {
-	b.values = make([][]interface{}, 0, b.size)
-}
-
-func (b *BatchInserter) flush() {
-	if len(b.values) > 0 {
-		clone := b.inserter.clone()
-		for i := range b.values {
-			clone.Values(b.values[i]...)
+func (b *BatchInserter) Next(dst interface{}) bool {
+	clone := b.inserter.clone()
+	i := 0
+	for value := range b.Values {
+		i++
+		clone.Values(value)
+		if b.size == i {
+			break
 		}
-		b.next <- clone
-		b.reset()
 	}
-}
-
-// Values pushes a value to be inserted as part of the batch.
-func (b *BatchInserter) Values(values ...interface{}) *BatchInserter {
-	b.mu.Lock()
-	defer b.mu.Unlock()
-
-	b.values = append(b.values, values)
-	if len(b.values) >= b.size {
-		b.flush()
+	if i == 0 {
+		return false
 	}
-	return b
+	b.err = clone.Iterator().All(dst)
+	return (b.err == nil)
 }
 
-// Next returns a channel that receives new q elements.
-func (b *BatchInserter) Next() chan Inserter {
-	b.mu.Lock()
-	defer b.mu.Unlock()
-
-	return b.next
+func (b *BatchInserter) Exec() error {
```

Comment:
I like this, this is like waiting for all batches to be executed and ignoring the results if they're not errors.